### PR TITLE
Introduce ExternalTypeRefInfoProvider

### PR DESCRIPTION
**Pull request by augusto2112**
Introduce ExternalTypeRefInfoProvider

LLDB would like to cache typeref information to accelerate finding type
information. This patch adds an optional interface that allows of
typeref to register and provider field descriptor information for faster
lookups.

### CODE REVIEW

File: include/swift/Reflection/ReflectionContext.h [hunk 1]
```diff
@@ -772,7 +772,7 @@ class ReflectionContext
   /// \return
```
Comment (adrian-prantl) [inline include/swift/Reflection/ReflectionContext.h:772]: comment is now out of date

File: include/swift/Remote/TypeInfoProvider.h [hunk 2]
```diff
@@ -32,6 +45,26 @@ struct TypeInfoProvider {
...
+
+  /// Register the field descriptors of a reflection info with a given id with
+  /// their corresponding mangled names. The amount of field descriptors and
+  /// mangled names should be the same. If a field descriptor does not have a
+  /// mangled name a corresponding empty string should be in the mangled_names
+  /// vector.
+  virtual void registerFieldDescriptors(
+      uint64_t InfoID,
+      const swift::reflection::FieldSection &FieldDescriptors,
+      const std::vector<std::string> &MangledNames) = 0;
```
Comment (adrian-prantl) [inline include/swift/Remote/TypeInfoProvider.h:57]: `const std::vector` sounds like it could be `ArrayRef`?

File: include/swift/Remote/TypeInfoProvider.h [hunk 3]
```diff
@@ -32,6 +45,26 @@ struct TypeInfoProvider {
...
+
+  /// Retrieve a pair representing the reflection info id and the offset of a
+  /// field descriptor in the field section buffer, if available.
```
Comment (adrian-prantl) [inline include/swift/Remote/TypeInfoProvider.h:60]: maybe define a struct with descriptive field names for the pair?

File: stdlib/public/Reflection/TypeRefBuilder.cpp [hunk 4]
```diff
@@ -263,23 +316,32 @@ RemoteRef<FieldDescriptor> TypeRefBuilder::getFieldTypeInfo(const TypeRef *TR) {
...
   // don't any useful information on which image to look for the type.
   if (ModuleName && ModuleName != llvm::StringRef("__C"))
     for (size_t i = 0; i < ReflectionInfos.size(); ++i)
-      if (llvm::is_contained(ReflectionInfos[i].PotentialModuleNames,
-                             ModuleName))
-        if (auto FD = findFieldDescriptorAtIndex(i, *MangledName))
-          return *FD;
+      if ((!ExternalTypeInfo ||
+           ExternalTypeInfo->shouldParseReflectionInfoWithId(i)))
+        if (llvm::is_contained(ReflectionInfos[i].PotentialModuleNames,
```
Comment (adrian-prantl) [inline stdlib/public/Reflection/TypeRefBuilder.cpp:331]: Not sure if this is much better, bug have you considered a SmallSet for PotentialModuleNames (in the small case it's also a linear list).

File: include/swift/Remote/TypeInfoProvider.h [hunk 5]
```diff
@@ -32,6 +45,26 @@ struct TypeInfoProvider {
   /// info, for example.
   virtual const reflection::TypeInfo *
   getTypeInfo(llvm::StringRef mangledName) = 0;
+
+  /// Register the field descriptors of a reflection info with a given id with
+  /// their corresponding mangled names. The amount of field descriptors and
+  /// mangled names should be the same. If a field descriptor does not have a
```
Comment (adrian-prantl) [inline include/swift/Remote/TypeInfoProvider.h:51]: should or must?

File: stdlib/public/Reflection/TypeRefBuilder.cpp (hunk 4 above)
Comment (augusto2112) [reply stdlib/public/Reflection/TypeRefBuilder.cpp:331]: This unfortunately can't be a SmallSet because the stdlib doesn't ship with it. I'm not even sure a set would be that much faster since there are so few elements.

**Comment by augusto2112** [comment]
@swift-ci smoke test

File: include/swift/Reflection/TypeRefBuilder.h [hunk 6]
```diff
@@ -982,35 +991,46 @@ class TypeRefBuilder {
   IntVariableReader OpaqueIntVariableReader;
 
 public:
-  template<typename Runtime>
```
Comment (mikeash) [inline include/swift/Reflection/TypeRefBuilder.h:985]: Looks like this is just a formatting change. Can we do that as a separate PR if so?

**Comment by mikeash** [comment]
Neat idea. Would it make sense to roll this logic into Remote Mirror rather than having it provided externally, or is it going to be very specific to LLDB in some way?

It would also be good to have tests to exercise the case when an external provider is given.

File: include/swift/Reflection/TypeRefBuilder.h (hunk 6 above)
Comment (augusto2112) [reply include/swift/Reflection/TypeRefBuilder.h:985]: It got formatted because of the new `remote::ExternalTypeRefInfoProvider *provider = nullptr` parameter, but I guess it formatted the whole constructor 😅  I'll remove the non-related parts.

**Comment by augusto2112** [comment]
> Neat idea. Would it make sense to roll this logic into Remote Mirror rather than having it provided externally, or is it going to be very specific to LLDB in some way?

Could you explain what you're envisioning when you suggest rolling this logic into Remote Mirror? If it could benefit the runtime besides LLDB I'd be happy to do it. To give more context, LLDB's use case for this is to cache a map of mangled names to field descriptor locators (the field descriptor and the offset in the typeref buffer), and then provide them back in future runs. You can check the lldb patch here: https://github.com/apple/llvm-project/pull/5100. The interesting file to look at i s is [SwiftMetadataCache.h](https://github.com/apple/llvm-project/pull/5100/files#diff-a88003491c8e8284dcf602fe7b2ba220aaecbda06c5ad5dada987607e0efd702R2] and [SwiftMetadataCache.cpp](https://github.com/apple/llvm-project/pull/5100/files#diff-9b59346d1112105e1631ae380eeb8b8c4e8fd4ea677bf0ec12ede7f1ddeee735R1]. If you see a way this could benefit the runtime as well we could change the approach.

**Comment by augusto2112** [comment]
> It would also be good to have tests to exercise the case when an external provider is given.

I'll try to come up with a way of testing this.

**Comment by augusto2112** [comment]
https://github.com/apple/llvm-project/pull/5100

@swift-ci smoke test

**Comment by mikeash** [comment]
@augusto2112 and I discussed this some offline. Persistent caching wouldn't fit well on the Swift side. It might be feasible to put most of the caching here, with a get/set function that the client could use to persist it across runs. Then clients other than LLDB could use it pretty easily if they needed it. However, we don't have such a need yet, so we'll just keep that as a future possibility.

File: stdlib/public/Reflection/TypeRefBuilder.cpp [hunk 7]
```diff
@@ -263,23 +309,32 @@ RemoteRef<FieldDescriptor> TypeRefBuilder::getFieldTypeInfo(const TypeRef *TR) {
...
   // don't any useful information on which image to look for the type.
   if (ModuleName && ModuleName != llvm::StringRef("__C"))
     for (size_t i = 0; i < ReflectionInfos.size(); ++i)
-      if (llvm::is_contained(ReflectionInfos[i].PotentialModuleNames,
-                             ModuleName))
-        if (auto FD = findFieldDescriptorAtIndex(i, *MangledName))
-          return *FD;
+      if ((!ExternalProvider ||
+           ExternalProvider->shouldParseReflectionInfoWithId(i)))
+        if (llvm::is_contained(ReflectionInfos[i].PotentialModuleNames,
```
Comment (adrian-prantl) [inline stdlib/public/Reflection/TypeRefBuilder.cpp:324]: I think this could use another comment explaining the algorithm and the contract with ExternalProvider here. I'm not sure it's obvious to the reader that !shouldParseReflectionInfoWithId(i) means we already parsed it.

File: stdlib/public/Reflection/TypeRefBuilder.cpp [hunk 8]
```diff
@@ -263,23 +309,32 @@ RemoteRef<FieldDescriptor> TypeRefBuilder::getFieldTypeInfo(const TypeRef *TR) {
...
+                  findFieldDescriptorAtIndex(i, *MangledName))
+            return *FD;
 
   // On failure, fill out the cache, ReflectionInfo by ReflectionInfo,
   // until we find the field descriptor we're looking for.
   for (size_t i = 0; i < ReflectionInfos.size(); ++i)
-    if (auto FD = findFieldDescriptorAtIndex(i, *MangledName))
-      return *FD;
+    if ((!ExternalProvider ||
+         ExternalProvider->shouldParseReflectionInfoWithId(i)))
```
Comment (adrian-prantl) [inline stdlib/public/Reflection/TypeRefBuilder.cpp:334]: would
```
if (ExternalProvider && !ExternalProvider->shouldParseReflectionInfoWithId(i)))
  continuel
```
be easier to read?

Comment (adrian-prantl) [reply stdlib/public/Reflection/TypeRefBuilder.cpp:334]: "shouldParse" doesn't explain why. Should we name it "didCacheReflectionInfo(i)" or simething more descriptive?

File: include/swift/Remote/ExternalTypeRefInfoProvider.h [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+      const llvm::ArrayRef<std::string> &MangledNames) = 0;
+
+  /// Retrieve a pair representing the reflection info id and the offset of a
+  /// field descriptor in the field section buffer, if available.
+  virtual llvm::Optional<FieldDescriptorLocator>
+  getFieldDescriptorLocator(const std::string &Name) = 0;
+
+  /// Returns whether looking for a mangled name in reflection info with a given
+  /// id could possibly return information that couldn't be provided by the
+  /// external provider.
```
Comment (adrian-prantl) [inline include/swift/Remote/ExternalTypeRefInfoProvider.h:68]: What would be an example?

File: stdlib/public/Reflection/TypeRefBuilder.cpp (hunk 8 above)
Comment (augusto2112) [reply stdlib/public/Reflection/TypeRefBuilder.cpp:334]: I named the ExternalProvider in a deliberately not specific way just in case someone else wanted to implement the interface with a different strategy. I can change it to be more concrete though.

File: include/swift/Remote/ExternalTypeRefInfoProvider.h (hunk 9 above)
Comment (augusto2112) [reply include/swift/Remote/ExternalTypeRefInfoProvider.h:68]: If we have all of the field descriptors for that ReflectionInfo on the cache, and `getFieldDescriptorLocator` didn't find anything, then the field descriptor we're looking for must not be in that ReflectionInfo.

**Comment by augusto2112** [comment]
https://github.com/apple/llvm-project/pull/5100

@swift-ci smoke test

**Comment by augusto2112** [comment]
@adrian-prantl made the new interface closer to how it's actually implemented, which should make the logic easier to understand.

File: include/swift/Remote/ExternalTypeRefCacheProvider.h [hunk 10]
```diff
@@ -0,0 +1,74 @@
+//===--- ExternalTypeRefCacheProvider.h - Abstract access to external caches of
+//typeref ------*- C++ -*-===//
```
Comment (adrian-prantl) [inline include/swift/Remote/ExternalTypeRefCacheProvider.h:2]: The `ExternalTypeInfoProvider` is called so because it provides typeinfo records from the outside. Here, I think `ExternalTypeRefCache` alone would suffice (without the `provider)?

File: include/swift/Remote/ExternalTypeRefCacheProvider.h [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file declares an abstract interface for external cache providers of
+//  typeref information
```
Comment (adrian-prantl) [inline include/swift/Remote/ExternalTypeRefCacheProvider.h:15]: `.`

Also should this be a file-wide doxygen comment?

File: include/swift/Remote/ExternalTypeRefCacheProvider.h [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+namespace reflection {
+
+template <typename T>
+class ReflectionSection;
+class FieldDescriptorIterator;
+using FieldSection = ReflectionSection<FieldDescriptorIterator>;
+}
+
+namespace remote {
+/// A struct with the information requited to locate a specific field
```
Comment (adrian-prantl) [inline include/swift/Remote/ExternalTypeRefCacheProvider.h:38]: `required`

File: include/swift/Remote/ExternalTypeRefCacheProvider.h (hunk 10 above)
Comment (adrian-prantl) [reply include/swift/Remote/ExternalTypeRefCacheProvider.h:2]: Thinking about it ... do we even need the `External` in the type name? I.e.: we could pass around a `TypeRefCache *ExternalTypeRefCache`

File: include/swift/Remote/ExternalTypeRefCacheProvider.h [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+  /// Cache the field descriptors of a reflection info with a given id with
+  /// their corresponding mangled names. The amount of field descriptors and
+  /// mangled names must be the same. If a field descriptor does not have a
+  /// mangled name a corresponding empty string must be in the mangled_names
+  /// array.
+  virtual void cacheFieldDescriptors(
+      uint64_t InfoID,
+      const swift::reflection::FieldSection &FieldDescriptors,
+      const llvm::ArrayRef<std::string> &MangledNames) = 0;
```
Comment (adrian-prantl) [inline include/swift/Remote/ExternalTypeRefCacheProvider.h:60]: `llvm::ArrayRef` is meant to be used as a value type, no need for the `const&`

File: include/swift/Remote/ExternalTypeRefCacheProvider.h (hunk 10 above)
Comment (augusto2112) [reply include/swift/Remote/ExternalTypeRefCacheProvider.h:2]: I agree with removing "provider", I do think "external" in the type name makes it more obvious what it is, versus as the variable name.

**Comment by augusto2112** [comment]
https://github.com/apple/llvm-project/pull/5100

@swift-ci smoke test

**Comment by augusto2112** [comment]
https://github.com/apple/llvm-project/pull/5100

@swift-ci smoke test

**Comment by augusto2112** [comment]
https://github.com/apple/llvm-project/pull/5100

@swift-ci smoke test Linux

File: include/swift/Reflection/ReflectionContext.h [hunk 14]
```diff
@@ -228,7 +229,8 @@ class ReflectionContext
   }
 
   template <typename T>
-  bool readMachOSections(
+  llvm::Optional<uint64_t>
```
Comment (adrian-prantl) [inline include/swift/Reflection/ReflectionContext.h:232]: document what the return value is?

Comment (adrian-prantl) [reply include/swift/Reflection/ReflectionContext.h:232]: And, like in the other patch — can this be a `uint32_t`?

**Review by adrian-prantl — APPROVED** [review]
LGTM with inline comments addressed

**Comment by augusto2112** [comment]
https://github.com/apple/llvm-project/pull/5100

@swift-ci smoke test Linux

**Comment by augusto2112** [comment]
https://github.com/apple/llvm-project/pull/5100

@swift-ci smoke test macOS
